test/dfv_test.cpp

```diff
@@ -0,0 +1,101 @@
+//=======================================================================
+// Copyright 2022 Ralf Kohrt
+
+// Distributed under the Boost Software License, Version 1.0. (See
+// accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//=======================================================================
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/depth_first_search.hpp>
+#include <boost/core/lightweight_test.hpp>
+
+using namespace boost;
+
+// Set up the vertex names
+enum vertex_id_t { u, v, w, x, y, z, N };
+
+
+struct counting_dfs_visitor
+{
+  counting_dfs_visitor():
+    vertex_events(0),
+    seen_edges(0)
+  {}
+
+  template<typename Vertex, typename Graph>
+  void initialize_vertex(Vertex v, const Graph& g)
+  {
+    ++vertex_events;
+  }
+
+  template<typename Vertex, typename Graph>
+  void start_vertex(Vertex v, const Graph& g)
+  {
+    ++vertex_events;
+  }
+
+  template<typename Vertex, typename Graph>
+  void discover_vertex(Vertex v, const Graph& g)
+  {
+    ++vertex_events;
+  }
+
+  template<typename Edge, typename Graph>
+  void examine_edge(Edge e, const Graph& g)
+  {
+    ++seen_edges;
+  }
+
+  template<typename Edge, typename Graph>
+  void tree_edge(Edge e, const Graph& g)
+  {}
+
+  template<typename Edge, typename Graph>
+  void back_edge(Edge e, const Graph& g)
+  {}
+
+  template<typename Edge, typename Graph>
+  void forward_or_cross_edge(Edge e, const Graph& g)
+  {}
+
+  template<typename Vertex, typename Graph>
+  void finish_vertex(Vertex v, const Graph& g)
+  {
+    ++vertex_events;
+  }
+
+  size_t vertex_events;
+  size_t seen_edges;
```

Comment:
I would actually be tempted to count each distinct event separately, to get a complete picture of what happens.
This would also help catch any subtle bugs.